### ISSUE TITLE
21.03 breaks dataloading with multiple CPUs

### ISSUE THREAD

**Issue by GR4HAM** [catalyst-team/catalyst#1133]
## 🐛 Bug Report

I have the following setup: 
* A dataloader that uses multiple workers 
* A model placed on a single GPU

When running the code, I believe catalyst assumes that ddp is happening and places the data on multiple GPUs even though ddp was not set to True. This is a problem starting in version 21.03 (I confirmed that in 20.12 it is not a problem).

### How To Reproduce
Steps to reproduce the behavior: run the following code sample on a machine with multiple GPUs and observe that the model is placed on all GPUs.

#### Code sample
Note that I used a dummy dataset and model to illustrate the problem.
```python
import os
import math
import torch
from catalyst import dl

os.environ["CUDA_VISIBLE_DEVICES"] = "2"

BATCH_SIZE = 4

class MyIterableDataset(torch.utils.data.IterableDataset):
    def __init__(self, start, end):
        super(MyIterableDataset).__init__()
        assert end > start, "this example code only works with end >= start"
        self.start = start
        self.end = end
        self.length = end-start

    def __iter__(self):
        return (x for x in [i*torch.ones(2,requires_grad=False) for i in range(self.start, self.end)])

    def __len__(self):
        return self.length

def worker_init_fn(worker_id):
    worker_info = torch.utils.data.get_worker_info()
    dataset = worker_info.dataset  # the dataset copy in this worker process
    overall_start = dataset.start
    overall_end = dataset.end
    # configure the dataset to only process the split workload
    per_worker = int(math.ceil((overall_end - overall_start) / float(worker_info.num_workers)))
    worker_id = worker_info.id
    dataset.start = overall_start + worker_id * per_worker
    dataset.end = min(dataset.start + per_worker, overall_end)        

ds = MyIterableDataset(start=3, end=100)
multi_cpu_data_loader = torch.utils.data.DataLoader(ds, batch_size=BATCH_SIZE, num_workers=4, worker_init_fn=worker_init_fn)

loaders = {"train":multi_cpu_data_loader,"valid":multi_cpu_data_loader}
dummy_model = torch.nn.Linear(2,1)

optimizer = torch.optim.Adam(dummy_model.parameters(), lr=0.001)
criterion = torch.nn.L1Loss()
target = torch.empty((BATCH_SIZE,1), dtype=torch.long).random_(5).cuda()

class CustomRunner(dl.Runner):
   
    def handle_batch(self, batch):
        output = self.model(batch)        
        loss = criterion(output,target[:output.shape[0]])       
        self.batch_metrics.update({"loss": loss.clone().detach()})

        if self.is_train_loader:
            loss.backward()            
            self.optimizer.step()
            self.optimizer.zero_grad()

runner = CustomRunner()

# model training
runner.train(
    model=dummy_model,
    optimizer=optimizer,
    loaders=loaders,
    logdir="../logs",
    num_epochs=2,
    verbose=True
)
```

### Expected behavior
The model should only use GPU 2.

### Environment
Catalyst version: 21.03
PyTorch version: 1.7.1
Torchvision version 0.8.2
TensorFlow version: N/A
TensorBoardX version: 2.1
OS: Ubuntu 16.04.6
Python version: 3.7
CUDA runtime version: 10.2
Nvidia driver version: 460.32.03
```

### Checklist
- [x ] bug description.
- [x ] steps to reproduce.
- [x ] expected behavior.
- [x ] environment.
- [x] code sample / screenshots.
- [x] I know, that I could [join Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion.

**Comment by Scitator** [comment]
Sorry, but I could not reproduce the issue (I just tested it on a few machines).
Could you please check the following setup:
1. run in the first notebook cell
    ```python
    import os
    os.environ["CUDA_VISIBLE_DEVICES"] = "2"
    ```
2. run in the second notebook cell
    ```python
    import os
    import math
    import torch
    from catalyst import dl
    
    BATCH_SIZE = 4
    
    class MyIterableDataset(torch.utils.data.IterableDataset):
        def __init__(self, start, end):
            super(MyIterableDataset).__init__()
            assert end > start, "this example code only works with end >= start"
            self.start = start
            self.end = end
            self.length = end-start
    
        def __iter__(self):
            return (x for x in [i*torch.ones(2,requires_grad=False) for i in range(self.start, self.end)])
    
        def __len__(self):
            return self.length
    
    def worker_init_fn(worker_id):
        worker_info = torch.utils.data.get_worker_info()
        dataset = worker_info.dataset  # the dataset copy in this worker process
        overall_start = dataset.start
        overall_end = dataset.end
        # configure the dataset to only process the split workload
        per_worker = int(math.ceil((overall_end - overall_start) / float(worker_info.num_workers)))
        worker_id = worker_info.id
        dataset.start = overall_start + worker_id * per_worker
        dataset.end = min(dataset.start + per_worker, overall_end)        
    
    ds = MyIterableDataset(start=3, end=100)
    multi_cpu_data_loader = torch.utils.data.DataLoader(ds, batch_size=BATCH_SIZE, num_workers=4, worker_init_fn=worker_init_fn)
    
    loaders = {"train":multi_cpu_data_loader,"valid":multi_cpu_data_loader}
    dummy_model = torch.nn.Linear(2,1)
    
    optimizer = torch.optim.Adam(dummy_model.parameters(), lr=0.001)
    criterion = torch.nn.L1Loss()
    target = torch.empty((BATCH_SIZE,1), dtype=torch.long).random_(5).cuda()
    
    class CustomRunner(dl.Runner):
       
        def handle_batch(self, batch):
            output = self.model(batch)        
            loss = criterion(output,target[:output.shape[0]])       
            self.batch_metrics.update({"loss": loss.clone().detach()})
    
            if self.is_train_loader:
                loss.backward()            
                self.optimizer.step()
                self.optimizer.zero_grad()
    
    runner = CustomRunner()
    
    # model training
    runner.train(
        model=dummy_model,
        optimizer=optimizer,
        loaders=loaders,
        logdir="../logs",
        num_epochs=2,
        verbose=True
    )
    ```

I suppose running such code in one cell does not affect `CUDA_VISIBLE_DEVICES` properly.

If you are using `.py` scripts, could you please try running with `CUDA_VISIBLE_DEVICES=2 python run.py`?

PS. nice test case :+1:

**Comment by GR4HAM** [comment]
Hi, thank you for your quick response. Indeed what you suggest works. It is very strange why I don't have this problem with 20.12 but I can make the above test case work by simply setting the environment before importing torch as such:

```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
import math
import torch
from catalyst import dl
```
So, apologies, problem appears to be from my side. Thanks again.

**Comment by Scitator** [comment]
A lot of underhood CUDA stuff could be changed after dl frameworks imports ;)
